### PR TITLE
Add new Livestock Trough action

### PR DESCRIPTION
**Pull request by paulsimonandrews**
https://eaflood.atlassian.net/browse/FCEP-120

New action with OR condition that aligns nicely with ELM

### CODE REVIEW

File: server/rules-engine/helper.js [hunk 1]
```diff
@@ -1,5 +1,32 @@
...
+  LV7: {
+    // LV7 has different eligibility criteria as only one rule needs to be true
+    // so we override the default eligibility check with this function
+    // Logic is also a bit odd: if there are no eligibility rules enabled then
+    // we assume you are eligible, otherwise either of one of the two eligibilty rules
+    // must pass if enabled. It's not ideal but works for this limited use case.
+    eligibilityCheckOverride: (result, enabledRules) => {
+      const ruleNames = enabledRules.map((rule) => rule.event.type)
+      const passedRuleNames = result.events.map((event) => event.type)
+      return ruleNames.length > 0 ? ruleNames.some((rule) => passedRuleNames.includes(rule)) : true
```
Comment (djbingham) [inline server/rules-engine/helper.js:25]: If the water pollution, reintroduced grazing and SSSI rules are all enabled, this check return true for a parcel which **only** passes the SSSI rule. That's not correct as the requirement should be that one or both of the water pollution or grazing rules pass **in addition to** the SSSI rule.

File: test/services/parcelActionsService.test.js [hunk 2]
```diff
@@ -29,18 +29,17 @@ jest.mock('../../server/services/parcelService', () => {
...
 const parcelActionsService = require('../../server/services/parcelActionsService')
 
 describe('parcelActionService', () => {
   beforeEach(() => {
     jest.clearAllMocks()
-    rulesEngineService.doEligibilityRun.mockImplementation((r, p, e, successCallback) => {
-      successCallback()
-      return Promise.resolve({ failingRules: [] })
+    rulesEngineHelper.eligibilityRun.mockImplementation((action, parcelData, options) => {
+      return Promise.resolve({ failingRules: [], actionPassed: true })
```
Comment (djbingham) [inline test/services/parcelActionsService.test.js:42]: Not worth changing now but FYI, Jest has a few helper functions for simple mocks like this:
```
rulesEngineHelper.eligibilityRun.mockResolvedValue({ failingRules: [], actionPassed: true })
```
https://jestjs.io/docs/en/mock-function-api#mockfnmockresolvedvaluevalue

**Review by djbingham — APPROVED** [review]
As you're aware, the eligibility check override is not correct as it will pass if any single rule passes, whereas the desired behaviour is either of two specific rules must pass in addition to any other enabled rules. That nested logic should ideally be configured in the JSON action definition.

However, as the intended use of this prototype has reached its end, it's certainly not worth the effort to implement nested rule logic properly now. If we return to this prototype in future, it's likely the rule definitions would move out of the rules engine which may significantly change how complex rule combinations would be implemented anyway.

File: server/rules-engine/helper.js (hunk 1 above)
Comment (paulsimonandrews) [reply server/rules-engine/helper.js:25]: Agreed,  having thought about it for I while I realised there was no nice way of doing this properly in a sensible amount of time so went with a non-ideal solution for the reasons you previously sate about this being a time limited prototype. But yes it should be fixed going forward if this stuff is going to be reused.

File: test/services/parcelActionsService.test.js (hunk 2 above)
Comment (paulsimonandrews) [reply test/services/parcelActionsService.test.js:42]: Thanks.

**Comment by neilbmclaughlin** [comment]
Just an observation. I think there are large sections of code added here and in previous commits which would not be needed if a few small changes were made to the rules-engine-test package. In particular, if the rules engine returned a collection containing the rule description and detailed failure reason for all failed rules (this was tested as an approach in the draft PR https://github.com/neilbmclaughlin/rules-engine-test/pull/8 and required just a little more work to make it available in the package API).

There seemed to be a reluctance to change the package code and this is a definite downside to having the rules engine codebase outside this repo.

**Pull request merged**